src/conn/conn_page_history.c

```diff
@@ -0,0 +1,368 @@
+        __wt_cond_signal(session, page_history->report_cond);
+        WT_TRET(__wt_thread_join(session, &page_history->report_tid));
+        page_history->report_tid_set = false;
+    }
+
+    if (page_history->report_session != NULL) {
+        WT_TRET(__wt_session_close_internal(page_history->report_session));
+        page_history->report_session = NULL;
+    }
+
+    if (page_history->report_cond != NULL) {
+        __wt_cond_destroy(session, &page_history->report_cond);
+        page_history->report_cond = NULL;
+    }
+
+    __wt_hash_map_destroy(session, &page_history->pages);
+    return (ret);
+}
+
+/*
+ * __wt_conn_page_history_track_evict --
+ *     Track page eviction.
+ */
+int
+__wt_conn_page_history_track_evict(WT_SESSION_IMPL *session, WT_PAGE *page)
+{
+    WT_DECL_RET;
+    WT_PAGE_HISTORY *page_history;
+    WT_PAGE_HISTORY_ITEM *item;
+    WT_PAGE_HISTORY_KEY key;
+    uint64_t page_id;
+
+    page_history = &S2C(session)->page_history;
+    if (!page_history->enabled)
+        return (0);
+
+    /* So far this works only for disaggregated storage, as we don't have page IDs without it. */
+    if (!F_ISSET(S2BT(session), WT_BTREE_DISAGGREGATED))
+        return (0);
+
+    page_id = page->block_meta.page_id;
+    if (page_id == 0)
+        return (0);
+
+    item = NULL;
+    key.page_id = page_id;
+    key.table_id = S2BT(session)->id;
+
+    WT_ERR(__wt_hash_map_get(
+      session, page_history->pages, &key, sizeof(key), (void **)&item, NULL, true, true));
+
+    item->key = key;
+    item->page_type = page->type;
+
+    item->num_evicts++;
+
+err:
+    if (item != NULL)
+        __wt_hash_map_unlock(session, page_history->pages, &key, sizeof(key));
+    return (ret);
+}
+
+/*
+ * __wt_conn_page_history_track_read --
+ *     Track page read.
+ */
+int
+__wt_conn_page_history_track_read(WT_SESSION_IMPL *session, WT_PAGE *page)
+{
+    WT_DECL_RET;
+    WT_PAGE_HISTORY *page_history;
+    WT_PAGE_HISTORY_ITEM *item;
+    WT_PAGE_HISTORY_KEY key;
+    uint64_t current_ms, current_global_read_count, page_id;
+
+    page_history = &S2C(session)->page_history;
+    if (!page_history->enabled)
+        return (0);
+
+    current_global_read_count = __wt_atomic_add64(&page_history->global_read_count, 1);
+
+    /* So far this works only for disaggregated storage, as we don't have page IDs without it. */
+    if (!F_ISSET(S2BT(session), WT_BTREE_DISAGGREGATED))
+        return (0);
+
+    page_id = page->block_meta.page_id;
+    if (page_id == 0)
+        return (0);
+
+    item = NULL;
+    key.page_id = page_id;
+    key.table_id = S2BT(session)->id;
+
+    WT_ERR(__wt_hash_map_get(
+      session, page_history->pages, &key, sizeof(key), (void **)&item, NULL, true, true));
+
+    item->key = key;
+    item->page_type = page->type;
+
+    __wt_milliseconds(session, &current_ms);
+
+    if (item->last_read_timestamp != 0)
+        item->total_time_between_reads += current_ms - item->last_read_timestamp;
```

Comment:
What's the purpose of total_time_between_reads?  Since it adds up distances between adjacent reads, it amounts to the total time between the first read and the final read.  Is that useful?